OrbitLinuxTracing/UprobesReturnAddressManager.h

```diff
@@ -59,6 +61,75 @@ class UprobesReturnAddressManager {
     }
   }
 
+  // In case of callchain sampling we don't have the complete stack to patch,
+  // but only the callchain (as list of instruction pointers). In those,
+  // a uprobe address occurs instead of the caller of an instrumented function.
+  // This function patches the callchain, using the maps information to identify
+  // instruction pointers of uprobe code and using the return address safed in
+  // the uprobes.
+  bool PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t callchain_size,
+                      unwindstack::Maps* maps) {
+    if (!tid_uprobes_stacks_.contains(tid)) {
+      return true;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    std::vector<uint64_t> ips_to_patch;
+
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
+    //  hijacked an address at the same stack pointer.
+    for (uint64_t i = 0; i < callchain_size; i++) {
+      uint64_t ip = callchain[i];
+      unwindstack::MapInfo* map_info = maps->Find(ip);
+
+      if (map_info == nullptr || map_info->name != "[uprobes]") {
+        continue;
+      }
+
+      ips_to_patch.push_back(i);
+    }
+
+    // In case we already used all uprobes, we need to discard this sample.
+    // There are two situations where this may happen:
+    //  1. At the beginning of a capture, where we missed the first uprobes
+    //  2. When there are two many events, resulting in lost events or
+    //   processing out of order.
+    if (tid_uprobes_stack.size() < ips_to_patch.size()) {
+      ERROR(
+          "Discarding sample in an uprobe as some uprobe records are missing.");
```

Comment:
```suggestion
          "Discarding sample in a uprobe as some uprobe records are missing.");
```
This kind of "u" at the beginning of a word is a consonant sound.